parser/src/lexer.rs

```diff
@@ -494,6 +497,55 @@ where
         Ok(())
     }
 
+    /// Lex a single magic command.
+    fn lex_magic_command(&mut self, kind: MagicKind) -> LexResult {
+        let start_pos = self.get_pos();
+        for _ in 0..u32::from(kind.prefix_len()) {
+            self.next_char();
+        }
+        let mut value = String::new();
+        loop {
+            match self.window[0] {
+                Some('\\') => {
+                    // Only skip the line continuation if it is followed by a newline
+                    // otherwise it is a normal backslash which is part of the magic command:
+                    //
+                    //        Skip this backslash
+                    //        v
+                    //   !pwd \
+                    //      && ls -a | sed 's/^/\\    /'
+                    //                          ^^
+                    //                          Don't skip these backslashes
+                    if matches!(self.window[1], Some('\n' | '\r')) {
+                        self.next_char();
+                        self.next_char();
+                    }
+                }
+                Some('\n' | '\r') | None => {
+                    let end_pos = self.get_pos();
+                    return Ok((
+                        Tok::MagicCommand { kind, value },
+                        TextRange::new(start_pos, end_pos),
+                    ));
+                }
+                Some(_) => {}
+            }
+            value.push(self.next_char().unwrap());
+        }
+    }
+
+    fn lex_and_emit_magic_command(&mut self) -> Result<(), LexicalError> {
+        if let [Some(c1), Some(c2)] = self.window[..2] {
+            if let Ok(kind) =
+                MagicKind::try_from([c1, c2]).map_or_else(|_| MagicKind::try_from(c1), Ok)
+            {
+                let magic_command = self.lex_magic_command(kind)?;
```

Comment:
I think we need to perform at least 1 lookahead to assert lexing `%` or `%%` as magic commands (I assume it is only a magic command if it is directly followed by a valid identifier start token)?

Reply:
I don't think so. IPython parses them regardless:

```python
from IPython.core.inputtransformer2 import TransformerManager

t = TransformerManager()

t.transform_cell('%')
# "get_ipython().run_line_magic('', '')\n"

t.transform_cell('%%')
# "get_ipython().run_cell_magic('', '', '')\n"

t.transform_cell('/')
# '()\n'

t.transform_cell(';')
# '("")\n'

t.transform_cell(',')
# '("")\n'

t.transform_cell('!')
# "get_ipython().system('')\n"

t.transform_cell('!!')
# "get_ipython().getoutput('')\n"

t.transform_cell('?')
# 'get_ipython().show_usage()\n'

t.transform_cell('??')
# 'get_ipython().show_usage()\n'
```

_The code can be run as is in the IPython REPL_